### PR TITLE
adding import_export functions

### LINKED ISSUE #125: unittest `test_export.py` fails to find and read the provided csv file
In the branch `export_import_archive`, for testing the export functionality, I compare the produced csv file during the test with a provided csv file in the directory. But the tests fails to find the file. Since the test does not fail on my machine, I wonder what is the reason? I change the relative path to the absolute path, I thought it might be the issue, but it did not help. If someone has an idea on how to solve this issue, I would be grateful.

### PR DESCRIPTION
**Pull request by max-hassani**
Integrating import to/export from archive functionality to pyiron projects.

### CODE REVIEW

File: pyiron_base/archiving/export.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import os
+import numpy as np
+from shutil import copyfile
+from pyfileindex import PyFileIndex 
+from pyiron_base import Project
```
Comment (jan-janssen) [inline pyiron_base/archiving/export.py]: We have to be a bit careful with the import of the `Project` class as this functionality should be imported in the `Project` class. I would suggest just passing a `Project` object instance and then using `Project().open()`.

File: pyiron_base/archiving/import.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import os
+import pandas 
+import numpy as np
+from shutil import copytree
+from pyiron_base import Project
```
Comment (jan-janssen) [inline pyiron_base/archiving/import.py]: Same here

File: pyiron_base/archiving/export.py (hunk 1 above)
Comment (max-hassani) [reply pyiron_base/archiving/export.py]: Good point! I removed it, and used `Project.open(<path>)`. Thanks for pointing it out.

File: pyiron_base/archiving/import.py (hunk 2 above)
Comment (max-hassani) [reply pyiron_base/archiving/import.py]: Done!

**Comment by max-hassani** [comment]
@jan-janssen I have tested both import and export functions. Both work fine.

File: pyiron_base/archiving/import_archive.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        elif isinstance(masterid, int) or isinstance(masterid, float):
+            masterid = int(masterid)
+            masterid_lst.append(job_id_lst[masterid])
+    return masterid_lst
+
+def import_jobs(project_instance, directory_to_import_to, archive_directory, df):
+    # Copy HDF5 files
+    archive_name = getdir(path=archive_directory)
+    cwd = os.getcwd()
+    directory_to_import_to = directory_to_import_to.split("/")[1]
```
Comment (pmrv) [inline pyiron_base/archiving/import_archive.py]: Use `os.path.dirname` and `os.path.abspath` to make things a bit clearer and shorter.

Comment (max-hassani) [reply pyiron_base/archiving/import_archive.py]: Done!

Comment (pmrv) [reply pyiron_base/archiving/import_archive.py]: Great!

File: pyiron_base/archiving/export_archive.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
+import os
+import numpy as np
+from shutil import copyfile
+from pyfileindex import PyFileIndex 
+import tarfile
+from shutil import rmtree
```
Comment (jan-janssen) [inline pyiron_base/archiving/export_archive.py:6]:
```suggestion
from shutil import rmtree


```

File: pyiron_base/archiving/export_archive.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+def new_job_id(job_id, job_translate_dict):
+    if isinstance(job_id, float) and not np.isnan(job_id):
+        job_id = int(job_id)
+    if isinstance(job_id, int):
+        return job_translate_dict[job_id]
+    else:
+        return None
+
```
Comment (jan-janssen) [inline pyiron_base/archiving/export_archive.py:17]:
```suggestion


```

File: pyiron_base/archiving/export_archive.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+def getdir(path): 
+    path_base_name = os.path.basename(path)
+    if path_base_name == "":
+        return os.path.basename(os.path.dirname(path))
+    else: 
+        return path_base_name
+
```
Comment (jan-janssen) [inline pyiron_base/archiving/export_archive.py:25]:
```suggestion


```

File: pyiron_base/archiving/export_archive.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+def update_project(project_instance, directory_to_transfer, archive_directory, df):
+    directory_to_transfer = directory_to_transfer.split("/")[1]+'/'
+    pr_transfer = project_instance.open('.')
+    dir_name_transfer = getdir(path=directory_to_transfer)
+    dir_name_archive = getdir(path=archive_directory)
+    path_rel_lst = [os.path.relpath(p, pr_transfer.project_path) for p in df["project"].values]
+    return [os.path.join(dir_name_archive, dir_name_transfer, p) if p != "." else os.path.join(dir_name_archive, dir_name_transfer) for p in path_rel_lst]
+
```
Comment (jan-janssen) [inline pyiron_base/archiving/export_archive.py:46]:
```suggestion


```

File: pyiron_base/archiving/export_archive.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+def filter_function(file_name):
+    return '.h5' in file_name
+
```
Comment (jan-janssen) [inline pyiron_base/archiving/export_archive.py:49]:
```suggestion


```

File: pyiron_base/archiving/export_archive.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+def generate_list_of_directories(df_files, directory_to_transfer, archive_directory):
+    path_rel_lst = [os.path.relpath(d, directory_to_transfer) for d in df_files.dirname.unique()]
+    dir_name_transfer = getdir(path=directory_to_transfer)
+    return [os.path.join(archive_directory, dir_name_transfer, p) if p != "." else os.path.join(archive_directory, dir_name_transfer) for p in path_rel_lst]
+
```
Comment (jan-janssen) [inline pyiron_base/archiving/export_archive.py:66]:
```suggestion


```

File: pyiron_base/archiving/export_archive.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+def compress_dir(archive_directory):
+    arch_comp_name=archive_directory+".tar.gz"
+    tar = tarfile.open(arch_comp_name, "w:gz")
+    tar.add(archive_directory, arcname=archive_directory)
+    tar.close()
+    rmtree(archive_directory)
+
```
Comment (jan-janssen) [inline pyiron_base/archiving/export_archive.py:73]:
```suggestion


```

File: pyiron_base/archiving/export_archive.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    for d in dir_lst: 
+        os.makedirs(d, exist_ok=True)
+    
+    # Copy files 
+    dir_name_transfer = getdir(path=directory_to_transfer)
+    for f in df_files.path.values:
+        copyfile(f, os.path.join(archive_directory, dir_name_transfer, os.path.relpath(f, directory_to_transfer)))
+    if compressed:
+        compress_dir(archive_directory)
+
```
Comment (jan-janssen) [inline pyiron_base/archiving/export_archive.py:110]:
```suggestion


```

File: pyiron_base/archiving/import_archive.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
+import os
+import pandas 
+import numpy as np
+from shutil import copytree, rmtree
+import tarfile
+
```
Comment (jan-janssen) [inline pyiron_base/archiving/import_archive.py:8]:
```suggestion


```

File: pyiron_base/archiving/import_archive.py [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+def getdir(path): 
+    path_base_name = os.path.basename(path)
+    if path_base_name == "":
+        return os.path.basename(os.path.dirname(path))
+    else: 
+        return path_base_name
+
```
Comment (jan-janssen) [inline pyiron_base/archiving/import_archive.py:16]:
```suggestion


```

File: pyiron_base/archiving/import_archive.py [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+def update_id_lst(record_lst, job_id_lst):
+    masterid_lst = []
+    for masterid in record_lst:
+        if np.isnan(masterid):
+            masterid_lst.append(masterid)
+        elif isinstance(masterid, int) or isinstance(masterid, float):
+            masterid = int(masterid)
+            masterid_lst.append(job_id_lst[masterid])
+    return masterid_lst
+
```
Comment (jan-janssen) [inline pyiron_base/archiving/import_archive.py:27]:
```suggestion


```

File: pyiron_base/archiving/import_archive.py [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+def extract_archive(archive_directory):
+    fname = archive_directory+".tar.gz" 
+    tar = tarfile.open(fname, "r:gz")
+    tar.extractall()
+    tar.close()
+
```
Comment (jan-janssen) [inline pyiron_base/archiving/import_archive.py:34]:
```suggestion


```

File: pyiron_base/project/generic.py [hunk 16]
```diff
@@ -1458,3 +1460,13 @@ def _update_jobs_in_old_database_format(self, job_name):
             elif db_entry_in_old_format:
                 for entry in db_entry_in_old_format:
                     self.db.item_update({"project": self.project_path}, entry["id"])
+
+    def export_to_archive(self,destination_path,csv_file_name='export.csv', compress=True):
```
Comment (jan-janssen) [inline pyiron_base/project/generic.py]:
```suggestion
    def export_to_archive(self, destination_path, csv_file_name='export.csv', compress=True):
```

File: pyiron_base/project/generic.py [hunk 17]
```diff
@@ -1458,3 +1460,13 @@ def _update_jobs_in_old_database_format(self, job_name):
...
+        export_archive.copy_files_to_archive(self.project_path, destination_path,compressed=compress)
+        df = export_archive.export_database(self,self.project_path,destination_path)
+        df.to_csv(csv_file_name)
+
+    def import_from_archive(self,origin_path,csv_file_name='export.csv',compress=True):
```
Comment (jan-janssen) [inline pyiron_base/project/generic.py]:
```suggestion
    def import_from_archive(self, origin_path, csv_file_name='export.csv', compress=True):
```

**Review by jan-janssen — COMMENTED** [review]
Some more PEP8 suggestions

File: pyiron_base/project/generic.py [hunk 18]
```diff
@@ -1461,12 +1461,12 @@ def _update_jobs_in_old_database_format(self, job_name):
...
+    def export_to_archive(self, destination_path, csv_file_name='export.csv', compress=True):
         export_archive.copy_files_to_archive(self.project_path, destination_path,compressed=compress)
         df = export_archive.export_database(self,self.project_path,destination_path)
         df.to_csv(csv_file_name)
 
-    def import_from_archive(self,origin_path,csv_file_name='export.csv',compress=True):
+    def import_from_archive(self, origin_path, csv_file_name='export.csv', compress=True):
         csv_path = csv_file_name
         df = pandas.read_csv(csv_path, index_col=0)
         import_archive.import_jobs(self,self.project_path,archive_directory=origin_path, df=df, compressed=compress)
```
Comment (jan-janssen) [inline pyiron_base/project/generic.py]:
```suggestion
        import_archive.import_jobs(self, self.path, archive_directory=origin_path, df=df, compressed=compress)
```

File: pyiron_base/project/generic.py [hunk 19]
```diff
@@ -1461,12 +1461,12 @@ def _update_jobs_in_old_database_format(self, job_name):
                 for entry in db_entry_in_old_format:
                     self.db.item_update({"project": self.project_path}, entry["id"])
 
-    def export_to_archive(self,destination_path,csv_file_name='export.csv', compress=True):
+    def export_to_archive(self, destination_path, csv_file_name='export.csv', compress=True):
         export_archive.copy_files_to_archive(self.project_path, destination_path,compressed=compress)
         df = export_archive.export_database(self,self.project_path,destination_path)
```
Comment (jan-janssen) [inline pyiron_base/project/generic.py]:
```suggestion
        export_archive.copy_files_to_archive(self.path, destination_path, compressed=compress)
        df = export_archive.export_database(self, self.path, destination_path)
```

File: pyiron_base/project/generic.py (hunk 18 above)
Comment (max-hassani) [reply pyiron_base/project/generic.py]: An empty line were added, in the new commit, before the function.

File: pyiron_base/project/generic.py (hunk 19 above)
Comment (max-hassani) [reply pyiron_base/project/generic.py]: An empty line were added, in the new commit, before the function.

**Comment by max-hassani** [comment]
The name of the export and import functions were suggested to be changed in the pyiron technical meeting to "packing" and "unpacking". These new naming were done in the previous commit.

File: pyiron_base/project/generic.py (hunk 19 above)
Comment (jan-janssen) [reply pyiron_base/project/generic.py]: This is not able an empty line, it is about `Project().path` vs `Project().project_path`

File: pyiron_base/project/generic.py (hunk 18 above)
Comment (jan-janssen) [reply pyiron_base/project/generic.py]: This is not able an empty line, it is about `Project().path` vs `Project().project_path`

File: pyiron_base/project/generic.py (hunk 19 above)
Comment (max-hassani) [reply pyiron_base/project/generic.py]: Done!

File: pyiron_base/project/generic.py (hunk 18 above)
Comment (max-hassani) [reply pyiron_base/project/generic.py]: Done!

**Comment by jan-janssen** [comment]
Codacy complains about a couple of tailing white spaces https://app.codacy.com/gh/pyiron/pyiron_base/pullRequest?prid=6504450 in addition GitHub complains about a incompatibility with the current master branch which needs to be resolved. @muh-hassani Can you take a look at these?

File: pyiron_base/project/generic.py [hunk 20]
```diff
@@ -1458,3 +1460,34 @@ def _update_jobs_in_old_database_format(self, job_name):
             elif db_entry_in_old_format:
                 for entry in db_entry_in_old_format:
                     self.db.item_update({"project": self.project_path}, entry["id"])
+
+
+    def packing(self, destination_path, csv_file_name='export.csv', compress=True):
```
Comment (jan-janssen) [inline pyiron_base/project/generic.py:1489]: I guess the `destination_path` should allow a relative path, an absolute path as well as a project object. So we should test that the variable is converted correspondingly.

File: pyiron_base/project/generic.py [hunk 21]
```diff
@@ -1458,3 +1460,34 @@ def _update_jobs_in_old_database_format(self, job_name):
...
+        destination_path (str) gives the ralative path, in which the project folder is copied and the compressed
+        csv_file_name (str) is the name of the csv file used to store the project table.
+        compress (boolian), if true, the function will compress the destination_path to a tar.gz file 
+        """
+        export_archive.copy_files_to_archive(self.path, destination_path,compressed=compress)
+        df = export_archive.export_database(self,self.path,destination_path)
+        df.to_csv(csv_file_name)
+
+
+    def unpacking(self, origin_path, csv_file_name='export.csv', compress=True):
```
Comment (jan-janssen) [inline pyiron_base/project/generic.py:1507]: I guess the `origin_path` should allow a relative path, an absolute path as well as a project object. So we should test that the variable is converted correspondingly.

**Comment by jan-janssen** [comment]
Finally it would be great to add unit tests for this functionality to demonstrate the different use cases.

**Comment by max-hassani** [comment]
> 
> 
> Codacy complains about a couple of tailing white spaces https://app.codacy.com/gh/pyiron/pyiron_base/pullRequest?prid=6504450 in addition GitHub complains about a incompatibility with the current master branch which needs to be resolved. @muh-hassani Can you take a look at these?

I don't understand why codacy complains about `Unexpected keyword argument 'dirs_exist_ok' in function call` in [https://app.codacy.com/gh/pyiron/pyiron_base/file/50421740557/issues/source?bid=20868371&fileBranchId=21014415#l42](https://app.codacy.com/gh/pyiron/pyiron_base/file/50421740557/issues/source?bid=20868371&fileBranchId=21014415#l42)
As far as I know there is nothing wrong with using the argument `dirs_exist_ok`

**Comment by jan-janssen** [comment]
> I don't understand why codacy complains about `Unexpected keyword argument 'dirs_exist_ok' in function call` in https://app.codacy.com/gh/pyiron/pyiron_base/file/50421740557/issues/source?bid=20868371&fileBranchId=21014415#l42
> As far as I know there is nothing wrong with using the argument `dirs_exist_ok`

It is only available in Python 3.8 https://docs.python.org/3/whatsnew/3.8.html#shutil

File: pyiron_base/project/generic.py (hunk 20 above)
Comment (max-hassani) [reply pyiron_base/project/generic.py:1489]: I think with the addition of these new line, the `destination_path` can be a relative path, absolute path, or a Project() object.  
[https://github.com/pyiron/pyiron_base/blob/748f6e24949d973fea457c76159032901880727a/pyiron_base/archiving/export_archive.py#L73](https://github.com/pyiron/pyiron_base/blob/748f6e24949d973fea457c76159032901880727a/pyiron_base/archiving/export_archive.py#L73)

File: pyiron_base/project/generic.py (hunk 21 above)
Comment (max-hassani) [reply pyiron_base/project/generic.py:1507]: I think with the addition of these new line, the `origin_path` can be a relative path, absolute path, or a Project() object.
[https://github.com/pyiron/pyiron_base/blob/748f6e24949d973fea457c76159032901880727a/pyiron_base/archiving/import_archive.py#L36](https://github.com/pyiron/pyiron_base/blob/748f6e24949d973fea457c76159032901880727a/pyiron_base/archiving/import_archive.py#L36)

**Comment by max-hassani** [comment]
With the latest commit, an issue with empty fields in the job table, during the import, is fixed.
[https://github.com/pyiron/pyiron_base/blob/9cbcae0c6a74cafd915a5fc1d363eb28cd711116/pyiron_base/archiving/import_archive.py#L61](https://github.com/pyiron/pyiron_base/blob/9cbcae0c6a74cafd915a5fc1d363eb28cd711116/pyiron_base/archiving/import_archive.py#L61)

File: tests/archiving/test_export.py [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+        self.status.finished = True
+
+
+class TestPacking(unittest.TestCase):
+    
+    @classmethod
+    def setUpClass(cls):
+        cls.arch_path = 'archive_folder'
+        cls.arch_path_comp = cls.arch_path+'_comp'
+        cls.pr = Project('test')
```
Comment (jan-janssen) [inline tests/archiving/test_export.py:32]: Be careful with the project path, it is recommended to use a file dependent path https://github.com/pyiron/pyiron_base/blob/master/tests/job/test_script.py#L12

**Comment by jan-janssen** [comment]
```
======================================================================
23
['archive_folder/test']
24
ERROR: test_exportedCSV (archiving.test_export.TestPacking)
25
----------------------------------------------------------------------
26
Traceback (most recent call last):
27
  File "/home/runner/work/pyiron_base/pyiron_base/tests/archiving/test_export.py", line 52, in test_exportedCSV
28
    df_known = pd.read_csv(csv_file_path)
29
  File "/usr/share/miniconda/envs/test/lib/python3.8/site-packages/pandas/io/parsers.py", line 605, in read_csv
30
    return _read(filepath_or_buffer, kwds)
31
  File "/usr/share/miniconda/envs/test/lib/python3.8/site-packages/pandas/io/parsers.py", line 457, in _read
32
    parser = TextFileReader(filepath_or_buffer, **kwds)
33
  File "/usr/share/miniconda/envs/test/lib/python3.8/site-packages/pandas/io/parsers.py", line 814, in __init__
34
    self._engine = self._make_engine(self.engine)
35
  File "/usr/share/miniconda/envs/test/lib/python3.8/site-packages/pandas/io/parsers.py", line 1045, in _make_engine
36
    return mapping[engine](self.f, **self.options)  # type: ignore[call-arg]
37
  File "/usr/share/miniconda/envs/test/lib/python3.8/site-packages/pandas/io/parsers.py", line 1862, in __init__
38
    self._open_handles(src, kwds)
39
  File "/usr/share/miniconda/envs/test/lib/python3.8/site-packages/pandas/io/parsers.py", line 1357, in _open_handles
40
    self.handles = get_handle(
41
  File "/usr/share/miniconda/envs/test/lib/python3.8/site-packages/pandas/io/common.py", line 639, in get_handle
42
    handle = open(
43
FileNotFoundError: [Errno 2] No such file or directory: '/home/runner/work/pyiron_base/pyiron_base/exemplary.csv'
```

But the file is located in: 
`/home/runner/work/pyiron_base/pyiron_base/tests/archiving/exemplary.csv'`

File: tests/archiving/test_export.py (hunk 22 above)
Comment (max-hassani) [reply tests/archiving/test_export.py:32]: `cls.arch_path` and `cls.arch_path_comp` are created during the test, wherever the test is executed (I guess in this case the root of `pyiron_base`). The problem was with the one I stored for comparison in the test directory.

**Comment by jan-janssen** [comment]
Is it necessary to have static files for the tests? I would prefer the HDF5 file `tests/archiving/sample_hdf.h5` to be written during the test job and the `csv` file to be created using pandas rather than storing those files in the test directory. In addition if it is not possible or not efficient to generate these files on the fly please store them in https://github.com/pyiron/pyiron_base/tree/master/tests/static then it is clear for other developers as well. Inside the tests sub directory there should only be the python script containing the unit tests.

**Comment by max-hassani** [comment]
> Is it necessary to have static files for the tests? I would prefer the HDF5 file `tests/archiving/sample_hdf.h5` to be written during the test job and the `csv` file to be created using pandas rather than storing those files in the test directory. In addition if it is not possible or not efficient to generate these files on the fly please store them in https://github.com/pyiron/pyiron_base/tree/master/tests/static then it is clear for other developers as well. Inside the tests sub directory there should only be the python script containing the unit tests. 

I have removed the tests related to these files and the test creates similar files during the run. I have forgotten to remove the files. I will do it now.

File: tests/archiving/test_import.py [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+        df_import["hamversion"] = float(df_import["hamversion"])
+        df_original["hamversion"] = float(df_original["hamversion"])
+        assert_frame_equal(df_original,df_import)
+    def test_import_compressed(self):
+        path_original = self.pr.path
+        path_import = self.imp_pr.path
+        path_original = getdir(path_original)
+        path_import = getdir(path_import)
+        compare_obj = dircmp(path_original,path_import)
+        self.assertEqual(len(compare_obj.diff_files),0)
```
Comment (jan-janssen) [inline tests/archiving/test_import.py]:
```suggestion
        self.assertEqual(len(compare_obj.diff_files),0)

```

File: tests/archiving/test_import.py [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+        df_import = self.imp_pr.job_table()
+        df_import.dropna(inplace=True,axis=1)
+        df_original.dropna(inplace=True,axis=1)
+        df_import.drop('project',inplace=True,axis=1)
+        df_original.drop('project',inplace=True,axis=1)
+        df_import.drop('id',inplace=True,axis=1)
+        df_original.drop('id',inplace=True,axis=1)
+        df_import["hamversion"] = float(df_import["hamversion"])
+        df_original["hamversion"] = float(df_original["hamversion"])
+        assert_frame_equal(df_original,df_import)
```
Comment (jan-janssen) [inline tests/archiving/test_import.py]:
```suggestion
        assert_frame_equal(df_original,df_import)

```

File: tests/archiving/test_import.py [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+        cls.pr.packing(destination_path=cls.arch_dir_comp,compress=True)
+        cls.file_location = os.path.dirname(os.path.abspath(__file__)).replace(
+            "\\", "/"
+        )
+    def setUp(self):
+        self.imp_pr = Project('imported')
+        self.imp_pr.remove_jobs_silently(recursive=True)
+        self.imp_pr.unpacking(origin_path=self.arch_dir_comp,compress=True)
+    def tearDown(self):
+        self.imp_pr.remove_jobs_silently(recursive=True)
```
Comment (jan-janssen) [inline tests/archiving/test_import.py:45]:
```suggestion
        self.imp_pr.remove_jobs_silently(recursive=True)

```

File: tests/archiving/test_import.py [hunk 26]
```diff
@@ -0,0 +1,70 @@
...
+        cls.job = cls.pr.create_job(job_type=ToyJob, job_name='toy')
+        cls.job.run()
+        cls.pr.packing(destination_path=cls.arch_dir_comp,compress=True)
+        cls.file_location = os.path.dirname(os.path.abspath(__file__)).replace(
+            "\\", "/"
+        )
+    def setUp(self):
+        self.imp_pr = Project('imported')
+        self.imp_pr.remove_jobs_silently(recursive=True)
+        self.imp_pr.unpacking(origin_path=self.arch_dir_comp,compress=True)
```
Comment (jan-janssen) [inline tests/archiving/test_import.py]:
```suggestion
        self.imp_pr.unpacking(origin_path=self.arch_dir_comp,compress=True)

```

File: tests/archiving/test_import.py [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+        cls.arch_dir = 'archive_folder' ## this is used to create a folder/a compressed file, are not path
+        cls.arch_dir_comp = cls.arch_dir+'_comp' ## this is used to create a folder/a compressed file, are not path
+        cls.pr = Project('test')
+        cls.pr.remove_jobs_silently(recursive=True)
+        cls.job = cls.pr.create_job(job_type=ToyJob, job_name='toy')
+        cls.job.run()
+        cls.pr.packing(destination_path=cls.arch_dir_comp,compress=True)
+        cls.file_location = os.path.dirname(os.path.abspath(__file__)).replace(
+            "\\", "/"
+        )
```
Comment (jan-janssen) [inline tests/archiving/test_import.py:37]:
```suggestion
        )

```

File: tests/archiving/test_import.py [hunk 28]
```diff
@@ -0,0 +1,70 @@
...
+    def __init__(self, project, job_name):
+        """A toyjob to test export/import functionalities."""
+        super(ToyJob, self).__init__(project, job_name)
+        self.input['input_energy'] = 100
+    # This function is executed
+    def run_static(self):
+        with self.project_hdf5.open("output/generic") as h5out:
+            h5out["energy_tot"] = self.input["input_energy"]
+        self.status.finished = True
+
```
Comment (jan-janssen) [inline tests/archiving/test_import.py:21]:
```suggestion


```

File: tests/archiving/test_import.py [hunk 29]
```diff
@@ -0,0 +1,70 @@
+import os
+import unittest
+from pyiron_base import Project
+from pyiron_base.archiving.import_archive import getdir
+from pandas._testing import assert_frame_equal
+from filecmp import dircmp
+from pyiron_base import PythonTemplateJob
+
```
Comment (jan-janssen) [inline tests/archiving/test_import.py:8]:
```suggestion


```

File: tests/archiving/test_import.py [hunk 30]
```diff
@@ -0,0 +1,70 @@
...
+class ToyJob(PythonTemplateJob):
+    def __init__(self, project, job_name):
+        """A toyjob to test export/import functionalities."""
+        super(ToyJob, self).__init__(project, job_name)
+        self.input['input_energy'] = 100
```
Comment (jan-janssen) [inline tests/archiving/test_import.py:14]:
```suggestion
        self.input['input_energy'] = 100

```

File: tests/archiving/test_import.py [hunk 31]
```diff
@@ -0,0 +1,70 @@
...
+        self.pr.packing(destination_path=self.arch_dir,compress=False)
+        self.imp_pr.remove_jobs_silently(recursive=True)
+        self.imp_pr.unpacking(origin_path=self.arch_dir,compress=False)
+        path_original = self.pr.path
+        path_import = self.imp_pr.path
+        path_original = getdir(path_original)
+        path_import = getdir(path_import)
+        compare_obj = dircmp(path_original,path_import)
+        self.assertEqual(len(compare_obj.diff_files),0)
+
```
Comment (jan-janssen) [inline tests/archiving/test_import.py:92]:
```suggestion


```

File: tests/archiving/test_import.py [hunk 32]
```diff
@@ -0,0 +1,70 @@
...
+        path_original = self.pr.path
+        path_import = self.imp_pr.path
+        path_original = getdir(path_original)
+        path_import = getdir(path_import)
+        compare_obj = dircmp(path_original,path_import)
+        self.assertEqual(len(compare_obj.diff_files),0)
+    def test_import_uncompress(self):
+        self.pr.packing(destination_path=self.arch_dir,compress=False)
+        self.imp_pr.remove_jobs_silently(recursive=True)
+        self.imp_pr.unpacking(origin_path=self.arch_dir,compress=False)
```
Comment (jan-janssen) [inline tests/archiving/test_import.py]:
```suggestion
        self.imp_pr.unpacking(origin_path=self.arch_dir, compress=False)
```

**Comment by jan-janssen** [comment]
The tests pass, that is great but the code quality can still be improved. I made a few suggestion and I highly recommend to use pecharm or any other IDE and look for their PEP8 related recommendations.

**Comment by max-hassani** [comment]
> 
> 
> The tests pass, that is great but the code quality can still be improved. I made a few suggestion and I highly recommend to use pecharm or any other IDE and look for their PEP8 related recommendations.

Thanks for the comments, @jan-janssen. Actually I am using VScode, and I guess it is the cause of my pain with the issues from Codacy. I removed all the empty lines at last to avoid those issues. I will add the lines and then try PyCharm for future.

File: pyiron_base/archiving/import_archive.py [hunk 33]
```diff
@@ -0,0 +1,79 @@
...
+
+def import_jobs(project_instance, directory_to_import_to, archive_directory, df, compressed=True):
+    # Copy HDF5 files
+    if isinstance(archive_directory,str):#if the archive_directory is a path
+        archive_directory = os.path.basename(archive_directory)
+    elif hasattr(archive_directory,'path'):#if the archive_directory is a project
```
Comment (jan-janssen) [inline pyiron_base/archiving/import_archive.py]: Please use the `static_isinstance()` function to check the type of a class - an example can be found here: https://github.com/pyiron/pyiron_base/blob/9fce7ef80c9e58a2a4e109ced403c3c371d7c292/pyiron_base/job/generic.py#L1050

**Comment by jan-janssen** [comment]
> Thanks for the comments, @jan-janssen. Actually I am using VScode, and I guess it is the cause of my pain with the issues from Codacy. I removed all the empty lines at last to avoid those issues. I will add the lines and then try PyCharm for future.

There are still a lot of functions which miss the spaces between the attributes. If you want to you can use https://black.readthedocs.io/en/stable/ to automatically do the formatting otherwise please use Pycharm to fix the PEP8 issues.

File: pyiron_base/archiving/import_archive.py (hunk 33 above)
Comment (max-hassani) [reply pyiron_base/archiving/import_archive.py]: It is done!

File: pyiron_base/archiving/export_archive.py [hunk 34]
```diff
@@ -0,0 +1,158 @@
...
+def export_database(
+    project_instance, directory_to_transfer,
+    archive_directory
+):
+    # here we first check wether the archive directory is a path
+    # or a project object
+    if isinstance(archive_directory, str):
+        archive_directory = os.path.basename(archive_directory)
+    elif hasattr(archive_directory, 'path'):
```
Comment (jan-janssen) [inline pyiron_base/archiving/export_archive.py]: please use `static_isinstance` here as well

Comment (max-hassani) [reply pyiron_base/archiving/export_archive.py]: Done!

File: pyiron_base/archiving/export_archive.py [hunk 35]
```diff
@@ -0,0 +1,167 @@
...
+        compress_dir(archive_directory)
+
+
+def export_database(
+    project_instance, directory_to_transfer,
+    archive_directory
+):
+    # here we first check wether the archive directory is a path
+    # or a project object
+    if static_isinstance(
```
Comment (jan-janssen) [inline pyiron_base/archiving/export_archive.py]: Please use the regular isinstance function here, we should only use the static_isinstance function to replace the hasattr part.

File: pyiron_base/archiving/import_archive.py [hunk 36]
```diff
@@ -0,0 +1,111 @@
...
+    tar.close()
+
+
+def import_jobs(
+    project_instance, directory_to_import_to, archive_directory,
+    df, compressed=True
+):
+    # Copy HDF5 files
+    # if the archive_directory is a path(string)/name of the compressed file
+    if static_isinstance(
```
Comment (jan-janssen) [inline pyiron_base/archiving/import_archive.py]: Same, here please use the regular isinstance function

File: pyiron_base/archiving/export_archive.py [hunk 37]
```diff
@@ -0,0 +1,167 @@
...
+        obj=archive_directory,
+        obj_type='builtins.str'
+    ):
```
Comment (jan-janssen) [inline pyiron_base/archiving/export_archive.py]:
```suggestion
    if isinstance(archive_directory, str):
```

File: pyiron_base/archiving/import_archive.py [hunk 38]
```diff
@@ -0,0 +1,111 @@
...
+        obj=archive_directory,
+        obj_type='builtins.str'
+    ):
```
Comment (jan-janssen) [inline pyiron_base/archiving/import_archive.py]:
```suggestion
    if isinstance(archive_directory, str):
```

**Review by jan-janssen — APPROVED** [review]
Looks good to me

File: pyiron_base/archiving/import_archive.py (hunk 38 above)
Comment (max-hassani) [reply pyiron_base/archiving/import_archive.py]: Sorry, this weekend I could not work. Thanks for updating it yourself.

**Comment by pmrv** [comment]
Sorry for the late nitpick (just came back from vacation), but is there chance we can rename the top-level functions to `pack` and `unpack`?  I think imperatives read better for stuff like this.  I'm happy to do the change myself, unless there was some discussion that I missed on why that would be a bad idea.  If you think it's too late now or we have production code using this already, I'll also not be offended. ;)

Otherwise I'm really happy with the functionality!

**Comment by max-hassani** [comment]
> 
> 
> Sorry for the late nitpick (just came back from vacation), but is there chance we can rename the top-level functions to `pack` and `unpack`? I think imperatives read better for stuff like this. I'm happy to do the change myself, unless there was some discussion that I missed on why that would be a bad idea. If you think it's too late now or we have production code using this already, I'll also not be offended. ;)
> 
> Otherwise I'm really happy with the functionality!

I will do it. But what do you mean by top level function? Do you mean changing the name of the `import_archive.py` and `export_archive.py`?

**Comment by max-hassani** [comment]
> 
> 
> Sorry for the late nitpick (just came back from vacation), but is there chance we can rename the top-level functions to `pack` and `unpack`? I think imperatives read better for stuff like this. I'm happy to do the change myself, unless there was some discussion that I missed on why that would be a bad idea. If you think it's too late now or we have production code using this already, I'll also not be offended. ;)
> 
> Otherwise I'm really happy with the functionality!

Sorry that I merged the PR without your review. I just thought you did not have time to review it.

**Comment by pmrv** [comment]
> > Sorry for the late nitpick (just came back from vacation), but is there chance we can rename the top-level functions to `pack` and `unpack`? I think imperatives read better for stuff like this. I'm happy to do the change myself, unless there was some discussion that I missed on why that would be a bad idea. If you think it's too late now or we have production code using this already, I'll also not be offended. ;)
> > Otherwise I'm really happy with the functionality!
> 
> I will do it. But what do you mean by top level function? Do you mean changing the name of the `import_archive.py` and `export_archive.py`?

I mean these two functions

https://github.com/pyiron/pyiron_base/blob/748f6e24949d973fea457c76159032901880727a/pyiron_base/project/generic.py#L1470

https://github.com/pyiron/pyiron_base/blob/748f6e24949d973fea457c76159032901880727a/pyiron_base/project/generic.py#L1485

When you call them, it's now
```python
pr.packing(...)
```
but I'd think 
```python
pr.pack(...)
```
reads cleaner.  But again it's a purely cosmetic thing.

> Sorry that I merged the PR without your review. I just thought you did not have time to review it.

No, that's fair the PR had been here long enough. ;)

**Comment by max-hassani** [comment]
> ```python
> pr.packing(...)
> ```
> 
> but I'd think
> 
> ```python
> pr.pack(...)
> ```
> 
> reads cleaner. But again it's a purely cosmetic thing.

Yeah, you are right. Pack/unpack sound much better than packing and unpacking. I will fix them via another PR.

**Comment by pmrv** [comment]
> > ```python
> > pr.packing(...)
> > ```
> > 
> > 
> > but I'd think
> > ```python
> > pr.pack(...)
> > ```
> > 
> > 
> > reads cleaner. But again it's a purely cosmetic thing.
> 
> Yeah, you are right. Pack/unpack sound much better than packing and unpacking. I will fix them via another PR.

Thanks a lot!

**Pull request merged**